pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
+	}
+	jobTpl.Labels[qualifiedNamePrefix+"reason"] = reason.String()
+	jobTpl.Labels[qualifiedNamePrefix+"target_agent_name"] = agentName
+	jobTpl.Labels[qualifiedNamePrefix+"target_agent_namespace"] = agentNs
+
+	if jobTpl.Annotations == nil {
+		jobTpl.Annotations = make(map[string]string)
+	}
+	jobTpl.Annotations[qualifiedNamePrefix+"controller_name"] = r.podName
+	jobTpl.Annotations[qualifiedNamePrefix+"controller_namespace"] = r.podNamespace
+	if rate > 0 {
+		jobTpl.Annotations[qualifiedNamePrefix+"rate"] = strconv.FormatFloat(rate, 'f', 4, 64)
+	}
+
+	log.Debugf("jobTpl.Labels: %#v\n", jobTpl.Labels)
+	log.Debugf("jobTpl.Annotations: %#v\n", jobTpl.Annotations)
+
+	jobTpl.Spec.Template.Labels = jobTpl.Labels
+	jobTpl.Spec.Template.ObjectMeta.Annotations = jobTpl.Annotations
+
+	log.Debugf("jobTpl.Spec.Template.Labels: %#v\n", jobTpl.Spec.Template.Labels)
+	log.Debugf("jobTpl.Spec.Template.Annotations: %#v\n", jobTpl.Spec.Template.Annotations)
+
+	jobTpl.Spec.Template.ObjectMeta.Labels[qualifiedNamePrefix+"target_agent_name"] = agentName
+	if rate > 0 {
+		jobTpl.Spec.Template.ObjectMeta.Annotations[qualifiedNamePrefix+"rate"] = strconv.FormatFloat(rate, 'f', 4, 64)
+	}
+
+	c := r.ctrl.GetManager().GetClient()
+	if err := c.Create(ctx, &jobTpl); err != nil {
+		return errors.ErrK8sFailedToCreateJob(err)
+	}
+
+	return nil
+}
+
+func (r *rebalancer) genPodModels() (podModels map[string][]*model.Pod, err error) {
+	mpods, ok := r.podMetrics.Load().(map[string]mpod.Pod)
+	if !ok {
+		return nil, errors.ErrEmptyReconileResult("pod metrics")
+	}
+
+	pods, ok := r.pods.Load().([]pod.Pod)
+	if !ok {
+		return nil, errors.ErrEmptyReconileResult("pod")
+	}
+
+	podModels = make(map[string][]*model.Pod)
+	for _, p := range pods {
+		if _, ok := podModels[p.Namespace]; !ok {
+			podModels[p.Namespace] = make([]*model.Pod, 0)
+		}
+		if mpod, ok := mpods[p.Name]; ok {
+			podModels[p.Namespace] = append(podModels[p.Namespace], &model.Pod{
+				Name:        p.Name,
+				Namespace:   p.Namespace,
+				MemoryLimit: p.MemLimit,
+				MemoryUsage: mpod.Mem,
+			})
+		}
+	}
+
+	return
+}
+
+func (r *rebalancer) namespaceByJobs() (jobmap map[string][]job.Job, err error) {
+	jobs, ok := r.jobs.Load().([]job.Job)
+	if !ok {
+		return nil, errors.ErrEmptyReconileResult("job")
+	}
+
+	jobmap = make(map[string][]job.Job)
+	for _, j := range jobs {
+		if _, ok := jobmap[j.Namespace]; !ok {
+			jobmap[j.Namespace] = make([]job.Job, 0)
+		}
+		jobmap[j.Namespace] = append(jobmap[j.Namespace], j)
+	}
+
+	return
+}
+
+func (r *rebalancer) genJobTpl() (jobTpl *job.Job, err error) {
+	tmpl, ok := r.jobTemplate.Load().(string)
+	if !ok {
+		return nil, errors.ErrJobTemplateNotFound()
+	}
+	jobTpl = &job.Job{}
+	err = r.decoder.DecodeInto([]byte(tmpl), jobTpl)
+	if err != nil {
+		return nil, errors.ErrFailedToDecodeJobTemplate(err)
+	}
+	return
+}
+
+func getDecreasedPodNames(prev, cur []pod.Pod, ns string) (podNames []string) {
+	podNames = make([]string, 0)
```

Comment:
meaning less allocation....

Reply:
thank you.
I fixed: https://github.com/vdaas/vald/pull/1187/commits/62ef3049bc13163129644fa166106e5602638084